File: RegExpBuilder.py
import re

class RegExpBuilder:

    def __init__(self):
        self._literal = ''
        self._ignoreCase = False
        self._specialCharactersInsideCharacterClass = set(['\\', '^', '-', ']'])
        self._specialCharactersOutsideCharacterClass = set(['\\', '.', '^', '$', '*', '+', '?', '(', ')', '[', '{'])
        self._min = -1
        self._max = -1
        self._of = ''
        self._ofAny = False
        self._from = ''
        self._notFrom = ''
        self._like = ''
        self._behind = ''
        self._notBehind = ''
        self._either = ''
        self._reluctant = False
        self._capture = False

    def _clear(self):
        self._ignoreCase = False
        self._multiLine = False
        self._min = -1
        self._max = -1
        self._of = ''
        self._ofAny = False
        self._from = ''
        self._notFrom = ''
        self._like = ''
        self._behind = ''
        self._notBehind = ''
        self._either = ''
        self._reluctant = False
        self._capture = False

    def _flushState(self):
        if self._of != '' or self._ofAny or self._from != '' or self._notFrom != '' or self._like != '':
            captureLiteral = '' if self._capture else '?:'
            quantityLiteral = self._getQuantityLiteral()
            characterLiteral = self._getCharacterLiteral()
            reluctantLiteral = '?' if self._reluctant else ''
            behindLiteral = '(?=' + self._behind + ')' if self._behind != '' else ''
            notBehindLiteral = '(?!' + self._notBehind + ')' if self._notBehind != '' else ''
            self._literal += '(' + captureLiteral + '(?:' + characterLiteral + ')' + quantityLiteral + reluctantLiteral + ')' + behindLiteral + notBehindLiteral
            self._clear()
  
    def _getQuantityLiteral(self):
        if self._min != -1:
            if self._max != -1:
                return '{' + str(self._min) + ',' + str(self._max) + '}'
            return '{' + str(self._min) + ',}'
        return '{0,' + str(self._max) + '}'
  
    def _getCharacterLiteral(self):
        if self._of != '':
            return self._of
        if self._ofAny:
            return '.'
        if self._from != '':
            return '[' + self._from + ']'
        if self._notFrom != '':
            return '[^' + self._notFrom + ']'
        if self._like != '':
            return self._like
  
    def getLiteral(self):
        self._flushState()
        return self._literal
  
    def getRegExp(self):
        self._flushState()
        flags = 0
        if self._ignoreCase:
            flags = flags | re.IGNORECASE
        if self._multiLine:
            flags = flags | re.MULTILINE
        return re.compile(self._literal, flags)
  
    def ignoreCase(self):
        self._ignoreCase = True
        return self
  
    def multiLine(self):
        self._multiLine = True
        return self
  
    def start(self):
        self._literal += '(?:^)'
        return self
  
    def end(self):
        self._flushState()
        self._literal += '(?:$)'
        return self
  
    def eitherLike(self, r):
        self._flushState()
        self._either = r(RegExpBuilder()).getLiteral()
        return self

  
    def orLike(self, r):
        eitherLike = self._either
        orLike = r(RegExpBuilder()).getLiteral()
        self._literal += '(?:(?:' + eitherLike + ')|(?:' + orLike + '))'
        self._clear()
        return self

  
    def exactly(self, n):
        self._flushState()
        self._min = n
        self._max = n
        return self

  
    def min(self, n):
        self._flushState()
        self._min = n
        return self

  
    def max(self, n):
        self._flushState()
        self._max = n
        return self

  
    def of(self, s):
        self._of = self._escapeOutsideCharacterClass(s)
        return self

  
    def ofAny(self):
        self._ofAny = True
        return self

  
    def fromClass(self, s):
        self._from = self._escapeInsideCharacterClass(''.join(s))
        return self

  
    def notFromClass(self, s):
        self._notFrom = self._escapeInsideCharacterClass(''.join(s))
        return self

  
    def like(self, r):
        self._like = r(RegExpBuilder()).getLiteral()
        return self

  
    def reluctantly(self):
        self._reluctant = True
        return self

  
    def behind(self, r):
        self._behind = r(RegExpBuilder()).getLiteral()
        return self

  
    def notBehind(self, r):
        self._notBehind = r(RegExpBuilder()).getLiteral()
        return self

  
    def asCapturingGroup(self):
        self._capture = True
        return self

  
    def _escapeInsideCharacterClass(self, s):
        return self._escapeSpecialCharacters(s, self._specialCharactersInsideCharacterClass)


    def _escapeOutsideCharacterClass(self, s):
        return self._escapeSpecialCharacters(s, self._specialCharactersOutsideCharacterClass)

  
    def _escapeSpecialCharacters(self, s, specialCharacters):
        escapedString = ''
        for i in range(len(s)):
            character = s[i]
            if character in specialCharacters:
                escapedString += '\\' + character
            else:
                escapedString += character

        return escapedString

File: RegExpBuilder.java


import java.util.Arrays;
import java.util.HashSet;
import java.util.regex.Pattern;

public class RegExpBuilder {
	private StringBuffer _literal;
	private Boolean _ignoreCase;
	private Boolean _multiLine;
	private HashSet<Character> _specialCharactersInsideCharacterClass;
	private HashSet<Character> _specialCharactersOutsideCharacterClass;
	private StringBuffer _escapedString;
	private int _min;
	private int _max;
	private String _of;
	private Boolean _ofAny;
	private String _from;
	private String _notFrom;
	private String _like;
	private String _behind;
	private String _notBehind;
	private String _either;
	private Boolean _reluctant;
	private Boolean _capture;
  
	public RegExpBuilder() {
		_literal = new StringBuffer();
		_specialCharactersInsideCharacterClass = new HashSet<Character>(Arrays.asList(new Character[]{ '^', '-', ']' }));
		_specialCharactersOutsideCharacterClass = new HashSet<Character>(Arrays.asList(new Character[]{ '.', '^', '$', '*', '+', '?', '(', ')', '[', '{' }));
		_escapedString = new StringBuffer();
		_clear();
	}
  
	private void _clear() {
		_ignoreCase = false;
		_multiLine = false;
		_min = -1;
		_max = -1;
		_of = "";
		_ofAny = false;
		_from = "";
		_notFrom = "";
		_like = "";
		_behind = "";
		_notBehind = "";
		_either = "";
		_reluctant = false;
		_capture = false;
	}
  
  	private void _flushState() {
  		if (_of != "" || _ofAny || _from != "" || _notFrom != "" || _like != "") {
  			String captureLiteral = _capture ? "" : "?:";
  			String quantityLiteral = getQuantityLiteral();
  			String characterLiteral = getCharacterLiteral();
  			String reluctantLiteral = _reluctant ? "?" : "";
  			String behindLiteral = _behind != "" ? "(?=" + _behind + ")" : "";
  			String notBehindLiteral = _notBehind != "" ? "(?!" + _notBehind + ")" : "";
  			_literal.append("(" + captureLiteral + "(?:" + characterLiteral + ")" + quantityLiteral + reluctantLiteral + ")" + behindLiteral + notBehindLiteral);
  			_clear();
    	}
  	}
  
  	private String getQuantityLiteral() {
  		if (_min != -1) {
  			if (_max != -1) {
  				return "{" + _min + "," + _max + "}";
  			}
  			return "{" + _min + ",}";
  		}
  		return "{0," + _max + "}";
  	}
  
  	private String getCharacterLiteral() {
  		if (_of != "") {
  			return _of;
  		}
  		if (_ofAny) {
  			return ".";
  		}
  		if (_from != "") {
  			return "[" + _from + "]";
  		}
  		if (_notFrom != "") {
  			return "[^" + _notFrom + "]";
  		}
  		if (_like != "") {
  			return _like;
  		}
  		return "";
  	}
  
  	public String getLiteral() {
  		_flushState();
  		return _literal.toString();
  	}
  
  	public Pattern getRegExp() {
  		_flushState();
  		int flags = 0;
  		if (_ignoreCase) {
  			flags = flags | Pattern.CASE_INSENSITIVE;
  		}
  		if (_multiLine) {
  			flags = flags | Pattern.MULTILINE;
  		}
  		return Pattern.compile(_literal.toString(), flags);
  	}
  
  	public RegExpBuilder ignoreCase() {
  		_ignoreCase = true;
  		return this;
  	}
  
  	public RegExpBuilder multiLine() {
  		_multiLine = true;
  		return this;
  	}
  
  	public RegExpBuilder start() {
  		_literal.append("(?:^)");
  		return this;
  	}
  
  	public RegExpBuilder end() {
  		_flushState();
  		_literal.append("(?:$)");
  		return this;
  	}
  
  	public RegExpBuilder either(RegExpBuilder r) {
  		_flushState();
  		_either = r.getLiteral();
  		return this;
  	}
  
  	public RegExpBuilder or(RegExpBuilder r) {
  		String either = _either;
  		String or = r.getLiteral();
  		_literal.append("(?:(?:" + either + ")|(?:" + or + "))");
  		_clear();
  		return this;
  	}
  
  	public RegExpBuilder exactly(int n) {
  		_flushState();
  		_min = n;
  		_max = n;
  		return this;
  	}
  
  	public RegExpBuilder min(int n) {
  		_flushState();
  		_min = n;
  		return this;
  	}
  
  	public RegExpBuilder max(int n) {
  		_flushState();
  		_max = n;
  		return this;
  	}
  
  	public RegExpBuilder of(String s) {
  		_of = _escapeOutsideCharacterClass(s);
  		return this;
  	}
  
  	public RegExpBuilder ofAny() {
  		_ofAny = true;
  		return this;
  	}
  
  	public RegExpBuilder from(char[] s) {
  		_from = _escapeInsideCharacterClass(new String(s));
  		return this;
  	}
  
  	public RegExpBuilder notFrom(char[] s) {
  		_notFrom = _escapeInsideCharacterClass(new String(s));
  		return this;
  	}
  
  	public RegExpBuilder like(RegExpBuilder r) {
  		_like = r.getLiteral();
  		return this;
  	}
  
  	public RegExpBuilder reluctantly() {
  		_reluctant = true;
  		return this;
  	}
  
  	public RegExpBuilder behind(RegExpBuilder r) {
  		_behind = r.getLiteral();
  		return this;
  	}
  
  	public RegExpBuilder notBehind(RegExpBuilder r) {
  		_notBehind = r.getLiteral();
  		return this;
  	}
  
  	public RegExpBuilder asCapturingGroup() {
  		_capture = true;
  		return this;
  	}
  
  	private String _escapeInsideCharacterClass(String s) {
  		return _escapeSpecialCharacters(s, _specialCharactersInsideCharacterClass);
  	}

  	private String _escapeOutsideCharacterClass(String s) {
  		return _escapeSpecialCharacters(s, _specialCharactersOutsideCharacterClass);
  	}
  
  	private String _escapeSpecialCharacters(String s, HashSet<Character> specialCharacters) {
  		_escapedString = new StringBuffer();
  		for (int i = 0; i < s.length(); i++) {
  			char character = s.charAt(i);
  			if (specialCharacters.contains(character)) {
  				_escapedString.append("\\" + character);
  			}
  			else {
  				_escapedString.append(character);
  			}
  		}
  		return _escapedString.toString();
  	}
}
